File: Scala/compoundlab/src/main/scala/MallardDuck.scala


class MallardDuck(observable: Option[IQuackableObservable] = None) extends IQuackable {

	private val name = "Mallard Duck"
	private val _observable : IQuackableObservable = observable getOrElse new Observable(this)


	override def registerObserver(observer : IObserver) : Unit = {
		this._observable.registerObserver(observer)
	}

	override def notifyObservers: String = {
		this._observable.notifyObservers
	}

	override def quack : String = {
		notifyObservers
		"Mallard duck quacked"
	}

	override def getName: String = {
		this.name
	}
}

File: Java/compoundlab/src/main/java/com/bbd/abstractFactoryPattern/MallardDuck.java
package com.bbd.abstractFactoryPattern;

public class MallardDuck implements IQuackable
{
	private String _name = "Mallard Duck";
	private IQuackObservable _observable;

	MallardDuck()
	{
		this._observable = new Observable(this);
	}

	MallardDuck(IQuackObservable observable)
	{
		this._observable = observable;
	}

	public String quack()
	{
		notifyObservers();
		return ("Mallard Duck Quacks");
	}

	public void registerObserver(IObserver observer)
	{
		this._observable.registerObserver(observer);
	}

	public String notifyObservers()
	{
		return (this._observable.notifyObservers());
	}

	@Override
	public String getName() { return  this._name;}

}
